<a href="http://cocl.us/pytorch_link_top">
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/Pytochtop.png" width="750" alt="IBM Product " />
</a> 

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/cc-logo-square.png" width="200" alt="cognitiveclass.ai logo" />

<h1>Linear Regression 1D: Prediction</h1>

<h2>Table of Contents</h2>
<p>In this lab, we will  review how to make a prediction in several different ways by using PyTorch.</h2>
<ul>
    <li><a href="#Prediction">Prediction</a></li>
    <li><a href="#Linear">Class Linear</a></li>
    <li><a href="#Cust">Build Custom Modules</a></li>
</ul>
<p>Estimated Time Needed: <strong>15 min</strong></p>

<hr>

<h2>Preparation</h2>

The following are the libraries we are going to use for this lab.

In [1]:
# These are the libraries will be used for this lab.
import torch

<!--Empty Space for separating topics-->

<h2 id="Prediction">Prediction</h2>

Let us create the following expressions:

$b=-1,w=2$

$\hat{y}=-1+2x$

First, define the parameters:

In [6]:
# Define w = 2 and b = -1 for y = wx + b
w = torch.tensor(2., requires_grad=True)
b = torch.tensor(-1., requires_grad=True)

print("Weight: ", w.item())
print("bias: ", b.item())

Weight:  2.0
bias:  -1.0


Then, define the function <code>forward(x, w, b)</code> makes the prediction: 

In [7]:
# Create Function Forward(x) for prediction (Forward Propagation)
def forward(x):
    yhat = w*x + b
    return yhat

Let's make the following prediction at <i>x = 1</i>

$\hat{y}=-1+2x$

$\hat{y}=-1+2(1)$

In [10]:
# predict y = 2x - 1 at x = 1
x = torch.tensor([[1.]])
yhat = forward(x)
print("The Prediction: ", yhat)

The Prediction:  tensor([[1.]], grad_fn=<AddBackward0>)


<!--Empty Space for separating topics-->

Now, let us try to make the prediction for multiple inputs:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2.png" width="500" alt="Linear Regression Multiple Input Samples" />

Let us construct the <code>x</code> tensor first. Check the shape of <code>x</code>.

In [11]:
# Create x Tensor and check the shape of x tensor
x = torch.tensor([[1.0], [2.0]])
print("The shape of x: ", x.shape)

The shape of x:  torch.Size([2, 1])


Now make the prediction: 

In [14]:
# Make the prediction of y = 2x - 1 at x = [1, 2]
yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.],
        [3.]], grad_fn=<AddBackward0>)


The result is the same as what it is in the image above.

<!--Empty Space for separating topics-->

<h3>Practice</h3>

Make a prediction of the following <code>x</code> tensor using the <code>w</code> and <code>b</code> from above.

In [17]:
# Practice: Make a prediction of y = 2x - 1 at x = [[1.0], [2.0], [3.0]]
x = torch.tensor([[1.0], 
                  [2.0], 
                  [3.0]])
# Prediction
yhat = forward(x)
print("prediction: ", yhat)

prediction:  tensor([[1.],
        [3.],
        [5.]], grad_fn=<AddBackward0>)


Double-click <b>here</b> for the solution.

<!-- Your answer is below:
yhat = forward(x)
print("The prediction: ", yhat)
-->

<!--Empty Space for separating topics-->

<h2 id="Linear">Class Linear</h2>

The linear class can be used to make a prediction. We can also use the linear class to build more complex models. Let's import the module:

In [18]:
# Import Class Linear
from torch.nn import Linear

Set the random seed because the parameters are randomly initialized:

In [46]:
# Set random seed
torch.manual_seed(1)

<!--Empty Space for separating topics-->

Let us create the linear object by using the constructor. The parameters are randomly created. Let us print out to see what <i>w</i> and <i>b</i> are.

In [47]:
# Create Linear Regression Model, and print out the parameters
lr = Linear(in_features=1, out_features=1, bias=True)
print("nn Linear Parameters \n", list(lr.parameters()),end="\n\n")

print("Weight (w): ", lr.weight.item())
print("Bias (b): ",lr.bias.item())

nn Linear Parameters 
 [Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]

Weight (w):  0.5152631998062134
Bias (b):  -0.44137823581695557


This is equivalent to the following expression:  

$w=0.5153$,  $b=-0.44$

$\hat{y}=-0.44+0.5153x$

Now let us make a single prediction at <i>x = [[1.0]]</i>.

In [30]:
# Make the prediction at x = [[1.0]]
x = torch.tensor([[2.]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.5891]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->

Similarly, you can make multiple predictions:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2vector_function.png" width="500" alt="Linear Class Sample with Multiple Inputs" />

Use model <code>lr(x)</code> to predict the result.

In [55]:
# Create the prediction using linear model
x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739],
        [0.5891]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->

<h3>Practice</h3>

Make a prediction of the following <code>x</code> tensor using the linear regression model <code>lr</code>.

In [57]:
# Practice: Use the linear regression model object lr to make the prediction.
x = torch.tensor([[1.0],[2.0],[3.0]])
yhat = lr(x)
print("prediction: ", yhat)

prediction:  tensor([[0.0739],
        [0.5891],
        [1.1044]], grad_fn=<AddmmBackward>)


<br>

Double-click <b>here</b> for the solution.

<!-- Your answer is below:
x=torch.tensor([[1.0],[2.0],[3.0]])
yhat = lr(x)
print("The prediction: ", yhat)
-->

In [70]:
class GanjilGenap:
#     def __init__(self, angka):
#         self.angka = angka
    def CheckGanjil(self, angka):
        print("Bilangan Ganjil!")
    def CheckGenap(self, angka):
        print("Bilangan Genap!")
    def __call__(self, angka):
        if angka % 2 == 1:
            CheckGanjil(angka)
        elif angka % 2 == 0:
            CheckGenap(angka)
            
checkangka = GanjilGenap()
checkangka(8)

NameError: name 'CheckGenap' is not defined

<!--Empty Space for separating topics-->

<h2 id="Cust">Build Custom Modules</h2>

Now, let's build a custom module. We can make more complex models by using this method later on. 

First, import the following library.

In [ ]:
# Library for this section
from torch import nn

Now, let us define the class: 

In [ ]:
# Customize Linear Regression Class

class LR(nn.Module):
    
    # Constructor
    def __init__(self, input_size, output_size):
        
        # Inherit from parent
        super(LR, self).__init__()
        self.linear = nn.Linear(input_size, output_size)
    
    # Prediction function
    def forward(self, x):
        out = self.linear(x)
        return out

Create an object by using the constructor. Print out the parameters we get and the model.

In [ ]:
# Create the linear regression model. Print out the parameters.

lr = LR(1, 1)
print("The parameters: ", list(lr.parameters()))
print("Linear model: ", lr.linear)

<!--Empty Space for separating topics-->

Let us try to make a prediction of a single input sample.

In [ ]:
# Try our customize linear regression model with single input

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

<!--Empty Space for separating topics-->

Now, let us try another example with multiple samples.

In [ ]:
# Try our customize linear regression model with multiple input

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

<!--Empty Space for separating topics-->

<h3>Practice</h3>

Create an object <code>lr1</code> from the class we created before and make a prediction by using the following tensor: 

In [ ]:
# Practice: Use the LR class to create a model and make a prediction of the following tensor.

x = torch.tensor([[1.0], [2.0], [3.0]])

Double-click <b>here</b> for the solution.

<!-- Your answer is below:
x=torch.tensor([[1.0],[2.0],[3.0]])
lr1=LR(1,1)
yhat=lr(x)
yhat
-->

 <!-- Your answer is below:
x=torch.tensor([[1.0],[2.0],[3.0]])
lr1=LR(1,1)
yhat=lr1(x)
yhat
-->

<a href="http://cocl.us/pytorch_link_bottom">
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/notebook_bottom%20.png" width="750" alt="PyTorch Bottom" />
</a>

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Other contributors: <a href="https://www.linkedin.com/in/michelleccarey/">Michelle Carey</a>, <a href="www.linkedin.com/in/jiahui-mavis-zhou-a4537814a">Mavis Zhou</a> 

<hr>

Copyright &copy; 2018 <a href="cognitiveclass.ai?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu">cognitiveclass.ai</a>. This notebook and its source code are released under the terms of the <a href="https://bigdatauniversity.com/mit-license/">MIT License</a>.